druid/src/widget/images.rs

```diff
@@ -0,0 +1,259 @@
+    FitWidth,
+    None,
+    ScaleDown,
+}
+
+pub enum FloatVert {
+    Top,
+    Center,
+    Bottom,
+}
+
+pub enum FloatHos {
+    Left,
+    Center,
+    Right,
+}
+
+fn get_scale_offset(
+    pairent: Size,
+    fit_box: Size,
+    fit_type: &FillStrat,
+    _fhos: &FloatHos,
+    _fvert: &FloatVert,
+) -> (Point, Point) {
+    let scalex = pairent.width / fit_box.width;
+    let scaley = pairent.height / fit_box.height;
+
+    let scale: Point = match fit_type {
+        FillStrat::Contain => {
+            let scale = scalex.min(scaley);
+            Point { x: scale, y: scale }
+        }
+        FillStrat::Cover => {
+            let scale = scalex.max(scaley);
+            Point { x: scale, y: scale }
+        }
+        FillStrat::Fill => Point {
+            x: scalex,
+            y: scaley,
+        },
+        FillStrat::FitHeight => Point {
+            x: scaley,
+            y: scaley,
+        },
+        FillStrat::FitWidth => Point {
+            x: scalex,
+            y: scalex,
+        },
+        FillStrat::ScaleDown => {
+            let scale = scalex.min(scaley).min(1.0);
+            Point { x: scale, y: scale }
+        }
+        FillStrat::None => Point { x: 1.0, y: 1.0 },
+    };
+
+    let origin_x = (pairent.width - (fit_box.width * scale.x)) / 2.0;
+    let origin_y = (pairent.height - (fit_box.height * scale.y)) / 2.0;
+    let origin = Point::new(origin_x, origin_y);
+
+    (scale, origin)
+}
+
+/// A widget that renders a Images
+pub struct Image<T> {
+    image_data: ImageData,
+    phantom: PhantomData<T>,
+
+    vert: FloatVert,
+    hoz: FloatHos,
+    fill: FillStrat,
+}
+
+impl<T: Data> Image<T> {
+    /// Create an Image-drawing widget from ImageData.
+    ///
+    /// The Image will scale to fit its box constraints.
+    pub fn new(image_data: ImageData, fillstr: FillStrat) -> impl Widget<T> {
+        Image {
+            image_data,
+            phantom: Default::default(),
+            vert: FloatVert::Center,
+            hoz: FloatHos::Center,
+            fill: fillstr,
+        }
+    }
+
+    /// Measure the Image's size
+    #[allow(clippy::needless_return)]
+    fn get_size(&self) -> Size {
+        // Fix me
+        Size::new(
+            self.image_data.x_pixels as f64,
+            self.image_data.y_pixels as f64,
+        )
+    }
+
+    pub fn set_fill(&mut self, newfil: FillStrat) {
+        self.fill = newfil;
+    }
+}
+
+impl<T: Data> Widget<T> for Image<T> {
+    fn event(&mut self, _ctx: &mut EventCtx, _event: &Event, _data: &mut T, _env: &Env) {}
+
+    fn lifecycle(&mut self, _ctx: &mut LifeCycleCtx, _event: &LifeCycle, _data: &T, _env: &Env) {}
+
+    fn update(&mut self, _ctx: &mut UpdateCtx, _old_data: &T, _data: &T, _env: &Env) {}
+
+    fn layout(
+        &mut self,
+        _layout_ctx: &mut LayoutCtx,
+        bc: &BoxConstraints,
+        _data: &T,
+        _env: &Env,
+    ) -> Size {
+        bc.debug_check("Image");
+
+        if bc.is_width_bounded() {
```

Comment:
what's the thinking here?

Reply:
SVG has the same thing. It means "if our parent has a concrete size use that size, otherwise become the size of the image".

Reply:
gotcha, I read it as the inverse. :)

Reply:
I think this was copy and pasted from from SVG